packages/ejson/lib/src/encoding.dart

```diff
@@ -0,0 +1,223 @@
+import 'package:objectid/objectid.dart';
+import 'package:sane_uuid/uuid.dart';
+import 'package:type_plus/type_plus.dart';
+
+import 'types.dart';
+
+/// Custom encoders for specific types. Use `register` to add a custom encoder.
+var customEncoders = <Type, Function>{};
+
+/// Whether to use relaxed encoding or not, default is false
+var relaxed = false;
+
+@pragma('vm:prefer-inline')
+/// Converts [value] to EJson 
+/// 
+/// Throws [MissingEncoder] if no encoder is registered for [value]'s type.
+EJsonValue toEJson(Object? value) => _encodeAny(value);
+
+EJsonValue _encodeAny(Object? value) {
+  return switch (value) {
+    null => null,
+    bool b => _encodeBool(b),
+    DateTime d => _encodeDate(d),
+    Defined<dynamic> d => _encodeDefined(d),
+    double d => _encodeDouble(d),
+    int i => _encodeInt(i),
+    Key k => _encodeKey(k),
+    Uint8List b => _encodeBinary(b, '00'),
+    Iterable<dynamic> l => _encodeArray(l),
+    Map<dynamic, dynamic> m => _encodeDocument(m),
+    ObjectId o => _encodeObjectId(o),
+    String s => _encodeString(s),
+    Symbol s => _encodeSymbol(s),
+    Undefined<dynamic> u => _encodeUndefined(u),
+    Uuid u => _encodeUuid(u),
+    _ => _encodeCustom(value),
+  };
+}
+
+EJsonValue _encodeArray(Iterable<dynamic> items) => items.map((e) => toEJson(e)).toList();
+
+EJsonValue _encodeBool(bool value) => value;
+
+EJsonValue _encodeCustom(Object value) {
+  final encoder = customEncoders[value.runtimeType.base];
+  if (encoder == null) {
+    throw MissingEncoder(value);
+  }
+  return encoder(value);
+}
+
+EJsonValue _encodeDate(DateTime value) {
+  return switch (relaxed) {
+    true => {'\$date': value.toIso8601String()},
+    false => {
+        '\$date': {'\$numberLong': value.millisecondsSinceEpoch.toString()},
+      },
+  };
+}
+
+EJsonValue _encodeDefined(Defined<dynamic> defined) => toEJson(defined.value);
+
+EJsonValue _encodeDocument(Map<dynamic, dynamic> map) => map.map((k, v) => MapEntry(k, toEJson(v)));
+
+EJsonValue _encodeDouble(double value) {
+  if (value.isNaN) {
+    return {'\$numberDouble': 'NaN'};
+  }
+  return switch (value) {
+    double.infinity => {'\$numberDouble': 'Infinity'},
+    double.negativeInfinity => {'\$numberDouble': '-Infinity'},
+    _ => switch (relaxed) {
+        true => value,
+        false => {'\$numberDouble': '$value'},
+      }
+  };
+}
+
+EJsonValue _encodeInt(int value, {bool long = true}) {
+  return switch (relaxed) {
+    true => value,
+    false => {'\$number${long ? 'Long' : 'Int'}': '$value'},
+  };
+}
+
+EJsonValue _encodeKey(Key key) => {'\$${key.name}Key': 1};
+
+@pragma('vm:prefer-inline')
+EJsonValue _encodeString(String value) => value;
+
+EJsonValue _encodeSymbol(Symbol value) => {'\$symbol': value.name};
+
+EJsonValue _encodeUndefined(Undefined<dynamic> undefined) => {'\$undefined': 1};
+
+EJsonValue _encodeUuid(Uuid uuid) => _encodeBinary(uuid.bytes.asUint8List(), "04");
+
+EJsonValue _encodeBinary(Uint8List buffer, String subtype) => {
```

Comment:
Might be a good idea to have the `subtype` as a named arg. We can also default to `00` as that would be the default binary encoding in MongoDB.

Reply:
Done. I have made the argument required, though.